File: src/main/java/net/gegy1000/terrarium/server/map/GenerationRegion.java
package net.gegy1000.terrarium.server.map;

public class GenerationRegion {
    public static final int BUFFER = 16;

    public static final int SIZE = 256;
    public static final int BUFFERED_SIZE = SIZE + BUFFER * 2;

    private final RegionTilePos pos;
    private final RegionData data;
    private final int minX;
    private final int minZ;

    public GenerationRegion(RegionTilePos pos, RegionData data) {
        this.pos = pos;
        this.data = data;
        this.minX = pos.getMinX() - GenerationRegion.BUFFER;
        this.minZ = pos.getMinZ() - GenerationRegion.BUFFER;
    }

    public RegionTilePos getPos() {
        return this.pos;
    }

    public int getMinX() {
        return this.minX;
    }

    public int getMinZ() {
        return this.minZ;
    }

    public RegionData getData() {
        return this.data;
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/map/GenerationRegion.kt
package net.gegy1000.terrarium.server.map

import net.gegy1000.terrarium.server.map.glob.GlobType
import net.gegy1000.terrarium.server.util.Coordinate
import net.minecraft.util.math.MathHelper

class GenerationRegion(
        val position: RegionTilePos,
        private val minimumCoordinate: Coordinate,
        private val scaledSize: Int,
        private val heights: ShortArray,
        private val globcover: Array<GlobType>
) {
    companion object {
        const val SIZE = 128
        const val SAMPLE_SIZE = SIZE + 1
    }

    fun getHeight(scaledX: Int, scaledZ: Int): Int {
        val localX = scaledX - MathHelper.floor(minimumCoordinate.blockX)
        val localZ = scaledZ - MathHelper.floor(minimumCoordinate.blockZ)
        return heights[localX + localZ * scaledSize].toInt()
    }

    fun getGlobType(scaledX: Int, scaledZ: Int): GlobType {
        val localX = scaledX - MathHelper.floor(minimumCoordinate.blockX)
        val localZ = scaledZ - MathHelper.floor(minimumCoordinate.blockZ)
        return globcover[localX + localZ * scaledSize]
    }
}
